VIcmd (or Screen) mode

t4nkw4rt opened the pull request:
Issue: If using FrankenWM remotely, I could not use ALT4 key.

Solution: I start a new command using CTRL-A, then optional numbers for how many times to repeat the command, and then the command itself: For Example:
CTRL-A 5 SHIFT-Return starts 5 Terminals

Press ESC or CTRL-A again to cancel the command.

Because 0-9 do not translate well over RDP, too, I changed these to F1-F10

Poweruser feature: If Person wants to hide the command window, set SHOW_COMMANDWIN to False. (Do this on your own risk, because there will be no indication, FrankenWM waits for input)

TODO: The command window itself looks pretty bad ugly.
It needs some more gfx care for sure.

t4nkw4rt commented:
Reason was because former version ended up so big, earlier this morning I merged the fullscreen-regression fix and git imported the entire history with it.  (I'm learning something new about git every day :)

sulami commented:
First of all:

```
frankenwm build options:
CFLAGS   = -std=c99 -pedantic -Wall -Wextra -I. -I/usr/local/include -I/usr/X11R6/include  -g
LDFLAGS  = -lc -lX11 -lxcb-util -lxcb-icccm -lxcb-keysyms -lxcb-ewmh -lxcb -g
CC       = cc
CC frankenwm.c
CC -o frankenwm
~/build/frankenwm:master λ cp config.def.h config.h 
~/build/frankenwm:t4nkw4rt/VIcmds λ make
frankenwm build options:
CFLAGS   = -std=c99 -pedantic -Wall -Wextra -I. -I/usr/local/include -I/usr/X11R6/include  -g
LDFLAGS  = -lc -lX11 -lxcb-util -lxcb-icccm -lxcb-keysyms -lxcb-ewmh -lxcb -g
CC       = cc
CC frankenwm.c
In file included from frankenwm.c:322:0:
config.h:154:26: error: ‘XK_F0’ undeclared here (not in a function)
        DESKTOPCHANGE(    XK_F0,                            9)
                          ^
config.h:75:26: note: in definition of macro ‘DESKTOPCHANGE’
     {  MOD4,             K,              change_desktop, {.i = N}}, \
                          ^
config.h:154:8: warning: missing initializer for field ‘arg’ of ‘key {aka struct <anonymous>}’ [-Wmissing-field-initializers]
        DESKTOPCHANGE(    XK_F0,                            9)
        ^
frankenwm.c:115:15: note: ‘arg’ declared here
     const Arg arg;
               ^
In file included from frankenwm.c:322:0:
config.h:154:8: warning: missing initializer for field ‘arg’ of ‘key {aka struct <anonymous>}’ [-Wmissing-field-initializers]
        DESKTOPCHANGE(    XK_F0,                            9)
        ^
frankenwm.c:115:15: note: ‘arg’ declared here
     const Arg arg;
               ^
In file included from frankenwm.c:322:0:
config.h:156:5: warning: missing initializer for field ‘arg’ of ‘key {aka struct <anonymous>}’ [-Wmissing-field-initializers]
     {  MOD4,             XK_Tab,        last_desktop,      {NULL}},
     ^
frankenwm.c:115:15: note: ‘arg’ declared here
     const Arg arg;
               ^
In file included from frankenwm.c:322:0:
config.h:158:5: warning: missing initializer for field ‘arg’ of ‘key {aka struct <anonymous>}’ [-Wmissing-field-initializers]
     {  MOD4|SHIFT,       XK_h,          rotate,            {.i = -1}},
     ^
frankenwm.c:115:15: note: ‘arg’ declared here
     const Arg arg;
               ^
In file included from frankenwm.c:322:0:
config.h:159:5: warning: missing initializer for field ‘arg’ of ‘key {aka struct <anonymous>}’ [-Wmissing-field-initializers]
     {  MOD4|SHIFT,       XK_l,          rotate,            {.i = +1}},
     ^
frankenwm.c:115:15: note: ‘arg’ declared here
     const Arg arg;
               ^
In file included from frankenwm.c:322:0:
config.h:161:5: warning: missing initializer for field ‘arg’ of ‘key {aka struct <anonymous>}’ [-Wmissing-field-initializers]
     {  MOD4|CONTROL,     XK_h,          rotate_client,     {.i = -1}},
     ^
frankenwm.c:115:15: note: ‘arg’ declared here
     const Arg arg;
               ^
In file included from frankenwm.c:322:0:
config.h:162:5: warning: missing initializer for field ‘arg’ of ‘key {aka struct <anonymous>}’ [-Wmissing-field-initializers]
     {  MOD4|CONTROL,     XK_l,          rotate_client,     {.i = +1}},
     ^
frankenwm.c:115:15: note: ‘arg’ declared here
     const Arg arg;
               ^
In file included from frankenwm.c:322:0:
config.h:164:5: warning: missing initializer for field ‘arg’ of ‘key {aka struct <anonymous>}’ [-Wmissing-field-initializers]
     {  MOD4|CONTROL|SHIFT, XK_h,        rotate_filled,     {.i = -1}},
     ^
frankenwm.c:115:15: note: ‘arg’ declared here
     const Arg arg;
               ^
In file included from frankenwm.c:322:0:
config.h:165:5: warning: missing initializer for field ‘arg’ of ‘key {aka struct <anonymous>}’ [-Wmissing-field-initializers]
     {  MOD4|CONTROL|SHIFT, XK_l,        rotate_filled,     {.i = +1}},
     ^
frankenwm.c:115:15: note: ‘arg’ declared here
     const Arg arg;
               ^
In file included from frankenwm.c:322:0:
config.h:168:5: warning: missing initializer for field ‘arg’ of ‘key {aka struct <anonymous>}’ [-Wmissing-field-initializers]
     {  MOD4,             XK_h,          resize_master,     {.i = -10}},
     ^
frankenwm.c:115:15: note: ‘arg’ declared here
     const Arg arg;
               ^
In file included from frankenwm.c:322:0:
config.h:169:5: warning: missing initializer for field ‘arg’ of ‘key {aka struct <anonymous>}’ [-Wmissing-field-initializers]
     {  MOD4,             XK_l,          resize_master,     {.i = +10}},
     ^
frankenwm.c:115:15: note: ‘arg’ declared here
     const Arg arg;
               ^
In file included from frankenwm.c:322:0:
config.h:170:5: warning: missing initializer for field ‘arg’ of ‘key {aka struct <anonymous>}’ [-Wmissing-field-initializers]
     {  MOD4,             XK_o,          resize_stack,      {.i = -10}},
     ^
frankenwm.c:115:15: note: ‘arg’ declared here
     const Arg arg;
               ^
In file included from frankenwm.c:322:0:
config.h:171:5: warning: missing initializer for field ‘arg’ of ‘key {aka struct <anonymous>}’ [-Wmissing-field-initializers]
     {  MOD4,             XK_p,          resize_stack,      {.i = +10}},
     ^
frankenwm.c:115:15: note: ‘arg’ declared here
     const Arg arg;
               ^
In file included from frankenwm.c:322:0:
config.h:174:5: warning: missing initializer for field ‘arg’ of ‘key {aka struct <anonymous>}’ [-Wmissing-field-initializers]
     {  MOD4|CONTROL,     XK_u,          adjust_borders,    {.i = -1}},
     ^
frankenwm.c:115:15: note: ‘arg’ declared here
     const Arg arg;
               ^
In file included from frankenwm.c:322:0:
config.h:175:5: warning: missing initializer for field ‘arg’ of ‘key {aka struct <anonymous>}’ [-Wmissing-field-initializers]
     {  MOD4|CONTROL,     XK_i,          adjust_borders,    {.i = +1}},
     ^
frankenwm.c:115:15: note: ‘arg’ declared here
     const Arg arg;
               ^
In file included from frankenwm.c:322:0:
config.h:177:5: warning: missing initializer for field ‘arg’ of ‘key {aka struct <anonymous>}’ [-Wmissing-field-initializers]
     {  MOD4|CONTROL,     XK_o,          adjust_gaps,       {.i = -1}},
     ^
frankenwm.c:115:15: note: ‘arg’ declared here
     const Arg arg;
               ^
In file included from frankenwm.c:322:0:
config.h:178:5: warning: missing initializer for field ‘arg’ of ‘key {aka struct <anonymous>}’ [-Wmissing-field-initializers]
     {  MOD4|CONTROL,     XK_p,          adjust_gaps,       {.i = +1}},
     ^
frankenwm.c:115:15: note: ‘arg’ declared here
     const Arg arg;
               ^
In file included from frankenwm.c:322:0:
config.h:180:5: warning: missing initializer for field ‘arg’ of ‘key {aka struct <anonymous>}’ [-Wmissing-field-initializers]
     {  MOD4|CONTROL,     XK_b,          togglepanel,       {NULL}},
     ^
frankenwm.c:115:15: note: ‘arg’ declared here
     const Arg arg;
               ^
In file included from frankenwm.c:322:0:
config.h:183:5: warning: missing initializer for field ‘arg’ of ‘key {aka struct <anonymous>}’ [-Wmissing-field-initializers]
     {  MOD4|CONTROL,     XK_q,          quit,              {.i = 0}},
     ^
frankenwm.c:115:15: note: ‘arg’ declared here
     const Arg arg;
               ^
frankenwm.c:3414:6: warning: ‘togglecmdmode’ defined but not used [-Wunused-function]
 void togglecmdmode()
      ^
Makefile:34: recipe for target 'frankenwm.o' failed
make: *** [frankenwm.o] Error 1
```

t4nkw4rt commented:
I am deeply sorry for that bad experience.

To me, it smells like XK_F0 (and propably (XK_F1-XK_F9) is not defined.

Here what happens on my side:
[user@flatant FrankenWM]$ git co master
Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.
[user@flatant FrankenWM]$ git co VIcmds
Switched to branch 'VIcmds'
Your branch is up-to-date with 'origin/VIcmds'.
[user@flatant FrankenWM]$ make clean
cleaning
removed 'frankenwm'
removed 'frankenwm.o'
[user@flatant FrankenWM]$ make
frankenwm build options:
CFLAGS   = -std=c99 -pedantic -Wall -Wextra -I. -I/usr/local/include -I/usr/X11R6/include  -g
LDFLAGS  = -lc -lX11 -lxcb-util -lxcb-icccm -lxcb-keysyms -lxcb-ewmh -lxcb -g
CC       = cc
CC frankenwm.c
CC -o frankenwm
[user@flatant FrankenWM]$

t4nkw4rt commented:
could you try this option in your Makefile?
-Wfatal-errors

Description:
     -Wfatal-errors
          This option causes the compiler to abort compilation on the first
          error occurred rather than trying to keep going and printing fur‐
          ther error messages.

Thanks.

sulami commented:
I found the issue, you accidentely defined `F0`, which screws up everything else
after that, no big deal. I fixed that and published a new upstream branch here
on GH to work on. `togglecmdmode` is still unused, and before this can make it
into master, there also needs to be documentation.

On Thu, Mar 31, 2016 at 08:09:19AM -0700, Ulrich Kaufmann wrote:
> I am deeply sorry for that bad experience.
> 
> To me, it smells like XK_F0 (and propably (XK_F1-XK_F9) is not defined.
> 
> Here what happens on my side:
> [user@flatant FrankenWM]$ git co master
> Switched to branch 'master'
> Your branch is up-to-date with 'origin/master'.
> [user@flatant FrankenWM]$ git co VIcmds
> Switched to branch 'VIcmds'
> Your branch is up-to-date with 'origin/VIcmds'.
> [user@flatant FrankenWM]$ make clean
> cleaning
> removed 'frankenwm'
> removed 'frankenwm.o'
> [user@flatant FrankenWM]$ make
> frankenwm build options:
> CFLAGS   = -std=c99 -pedantic -Wall -Wextra -I. -I/usr/local/include -I/usr/X11R6/include  -g
> LDFLAGS  = -lc -lX11 -lxcb-util -lxcb-icccm -lxcb-keysyms -lxcb-ewmh -lxcb -g
> CC       = cc
> CC frankenwm.c
> CC -o frankenwm
> [user@flatant FrankenWM]$ 
> 
> 
> ---
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/sulami/FrankenWM/pull/92#issuecomment-203979753

-- 

Robin Schroer

t4nkw4rt commented:
First thing this morning, I wanted to send you my own config.h, then saw your new comment pointing in the right direction.
Fixed the F10 binding.
and added the much needed Ctrl-A binding to the config.def.h file.

sulami commented:
Okay, I merged your latest additions and tweaked the new manpage section a bit.
One thing I am still thinking about is enabling this by default and binding it
to C-a, which might surprise some people that do not know of this functionality.
I think I like C-a as a default shortcut, even though it clashes with screen
(and many tmux configs).

I also have to test this somewhat thoroughly myself before actually pushing it
to master, so in this time, maybe we can come up with some way to not destroy
screen users' workflow by default. I have been thinking about just adding a
`#define` to disable this by default without burying it in the shortcuts section
of the config, but I am open for suggestions on how to handle this best.

t4nkw4rt commented:
Users who need tihs feature either use screen (ctrl-a) or tmux (ctrl-b), so they can bind this feature to the unused ctrl-binding (ctrl-a in my case, because I use tmux)

This feature should not clash with any existing command-binding, because all it does is replacing mod4 with ctrl-a and then use the same command-binding table. Only extra thing is the optional number on how many times to repeat the command.

Feel free to test it "auf Herz und Nieren" :)

THERE is one thing I am unsure about: can we use the normal CLEANMASK or should we use a new mask NOMOD4MASK. It works with this mask, but I do not know if we really need it.

Thanks a bunch.

I'm also open to any suggestions about a niecer looking command-window.
